Dev com xvalues

a-monsch opened the pull request:
introducing the feature, that if performing fits the center of mass (com) in each bin is used for the x-value, instead of the bin center.

This might simplify problems arising with the choice of the last bin edge.

 * [x] FF calculation
 * [x] Corrections

nshadskiy commented on helper/hooks_and_patches.py:
```diff
@@ -0,0 +1,347 @@
+                if len(histo_params) == 5:
+                    _, _, nbins, low, high = histo_params  # dont care about name and title
+                    bins = np.linspace(low, high, int(nbins) + 1)
+                elif len(histo_params) >= 4:
```
Should this be `>= 4`? This if statement is not orthogonal for `hist_params` of length 5.

a-monsch replied:
No, `==` would be right. Will be fixed in the upcomming commits.

nshadskiy commented on helper/hooks_and_patches.py:
```diff
@@ -0,0 +1,347 @@
+        if tot != 0:
+            _means.append((w1 * m1 + factor * w2 * m2) / tot)
+        else:
+            _means.append(0)
```
For my understanding, this happens when both histogram yields in a bin are 0, right? Does it make sense to set the mean to 0? As I understand, the mean is basically the x-value for each bin, so wouldn't it be better to set it to `(m1+m2)/2` or something?

a-monsch replied:
We could, or we could go with `float("nan")`?

From the computational perspective, no weights mean an empty bin at the end, so the value we place there might be arbitrarily chosen. When further down the calculation path an addition with non zero bin is performed the bin with no content will not contribute to the center of mass calculation.

nshadskiy replied:
True. I basically meant the bin center. If we use `nan` this will also result in a `nan` for an addition of an empty and an non-empty bin, I don't think we want that, right?

nshadskiy commented on configs/general_definitions.py:
```diff
@@ -9,14 +11,20 @@
+    binning: List[float],
+    hist: Union[ROOT.TH1, None] = None,
+) -> dict:
+    if hasattr(hist, _EXTRA_PARAM_FLAG) and getattr(hist, _EXTRA_PARAM_FLAG):
+        pass
```
Doesn't this mean that we again have a constant uncertainty for outside the bin range?

a-monsch replied:
Thanks for pointing this out, I accidentally added this part to the latest commit.

The Intention will be: If com-calculation is enabled, then we might not want to limit the fit within the histogram range, but within the com-values range (imagining a binning like `[0, 10, 20, 1000]`). If so, what would be an appropriate way to set it by the user?

nshadskiy replied:
I agree, I don't think that it's reasonable to use the current default if the last bin is very wide. Maybe we can use something like `com+2*left_unc_x` (for the last bin case). With left_unc_x I mean the distance to the lower bin edge of the last bin from the com value. Something similar for the first bin case but in the other direction.

nshadskiy commented on FF_calculation/FF_QCD.py:
```diff
@@ -361,7 +362,7 @@ def non_closure_correction(
     )
 
-    smoothed_graph, correction_dict = ff_func.smooth_function(
+    nominal_draw_obj, smoothed_graph, correction_dict = ff_func.smooth_function(
```
Add this also to the DR_SR correction. There also smooth_function is used.

a-monsch replied:
Done

nshadskiy commented on helper/ff_functions.py:
```diff
@@ -481,13 +508,70 @@ def get_yields_from_hists(
+def build_TGraph(
+    hist: Union[ROOT.TH1, None] = None,
+    return_components: bool = False,
+    convert_components_to_array: bool = True,
```
This parameter is not needed for the current use cases.

nshadskiy commented on helper/ff_functions.py:
```diff
@@ -735,7 +732,9 @@ def calculate_non_closure_correction_ttbar_fromMC(
 
 def smooth_function(
-    hist: Any, bin_edges: List[float], write_corrections: str
+    hist: Any, bin_edges: List[float],
```
Split in two lines

a-monsch replied:
Done

nshadskiy reviewed (approved):
Looks good to me! Can be merged, I guess.

nshadskiy commented on helper/ff_functions.py:
```diff
@@ -226,7 +226,11 @@ def apply_region_filters(
     if category_cuts is not None:
         for cut in category_cuts:
-            tmp[cut] = f"{cut} {category_cuts[cut]}"
+            if "#" not in category_cuts[cut]:
+                tmp[cut] = f"{cut} {category_cuts[cut]}"
+            else:
+                a, op, b = category_cuts[cut].split("#")
+                tmp[cut] = f"({cut} {a}) {op} ({cut} {b})"
```
Is there an example for that?

a-monsch replied:
I uploaded a verison of it in the configs, but I will extend the README i guess?

a-monsch commented:
Intermediary:

Additionally, to the stated points in the initial message, I added:

 * A simple cache, where in the case of sequentially applied non-closure corrections, if at one step in the chain the config of the individual correction changed, all further cached steps are ignored and a normal calculation (building a new cache) is performed.
If it is a non-closure correction for DR-SR correction, and the case described above appears, the DR_SR cached correction is actively removed. DR_SR corrections are also cached, where it checks for equality of all performed non-closure corrections within the scope of the DR_SR correction. In total, this might speed up the iteration time in the case of multiple corrections, where one doesn't need to redo the first correction each time a subsequent one is modified.
  * New option to non-closure corrections, previous "smoothed", "binwise" can be additionally combined into "binwise#[0,]+smoothed", "binwise#[-1,-2,]#+smoothed" or "binwise#[0,]#[-1,]+smoothed" where the nth-bin(s) is computed using binwise method and the rest in smoothed manner. For left bins, positive bin index is used. For last bins, the count is done using negative integers, starting at -1 (like in the example above). If both options are provided, then the binwise calculation is applied for the stated left and right bins.
  * Added a basic correction plotting function using Matplotlib and mplhep, since the upper modifications produced bugged display results using the root version.
  * Added logic for all split arguments in configs, to be stored in one object, extended this functionality to currently all arguments that make sense to be split.
  
A refinement of the README still needs to be done. Additionally, I want to highlight a problem encountered when implementing the cache functionality, noted [here](https://github.com/KIT-CMS/TauFakeFactors/pull/10/commits/bfc8a4a7ffc1fa7f0c47a7abe3596b564496f430)

nshadskiy reviewed (approved):
I think for this PR everything is complete. Additional features should be added with another PR.

nshadskiy commented on configs/general_definitions.py:
```diff
@@ -90,23 +90,23 @@ def get_default_bandwidth(
+    "tau_fakes": "#B9AC70",  # (185, 172, 112)
+    "data_ff": "#FFA90E",  # (255, 169, 14)
     "fit_graph_mc_sub": "#5790fc",
     "fit_graph_unc": "#e42536",
```
Since you want to use the `OFFICIAL_CMS_COLOR_PALLET`, I would drop this here.

nshadskiy commented on helper/plotting.py:
```diff
@@ -249,33 +302,58 @@ def plot_FFs(
+
+    recreation_summary = dict(
+        variable=variable,
+        # ff_ratio=ff_ratio,  # will be adjusted in the function
+        # uncertainties=uncertainties,  # will be adjusted in the function
+        era=era,
+        channel=channel,
+        process=process,
+        category=category,
+        # output_path=output_path,  # required when recreating the plot
+        # logger=logger,  # required when recreating the plot
+        draw_option=draw_option,
+        # save_data=save_data,  # not needed when recreating the plot
```
I understand why you commented out the parameters, but I would just remove them. If you want to recreate a plot this here will probably not be the place where you will look for the relevant function inputs.
Same comment for the other functions.

nshadskiy replied:
Yes, I know. I only meant to remove the parameters that are commented out. If you want to know which parameters you need you will either way first look at the function and not a this dict, right?

a-monsch replied:
Ah, ok. Yes, agreed.
